docs/arch/adr-001-use-lettuce-core-lib-instead-of-dropwizard-redis.md

```diff
@@ -0,0 +1,34 @@
+# ADR 001 - Use lettuce-core library instead of dropwizard redis
+
+## Context
+
+There was an [intention](https://payments-platform.atlassian.net/browse/PP-6343) to use the 
+[dropwizard-redis](https://github.com/dropwizard/dropwizard-redis) library. The advantages of using this library are:
+
+* Configuration
+* Client lifecycle management
+* Client health checks
+* Dropwizard Metrics integration
+
+The dropwizard-redis library's io.dropwizard.redis.RedisClientFactory class 
+[expects](https://github.com/dropwizard/dropwizard-redis/blob/master/src/main/java/io/dropwizard/redis/RedisClientFactory.java#L54) 
```

Comment:
it's not evident/apparent to me why this line of code indicates that Redis is required to be running at application startup. The link shows that `RedisClientFactory.build()` attempts to open a connection.

Digging further through the code I can see that the `RedisClientBundle` [builds the connection](https://github.com/dropwizard/dropwizard-redis/blob/master/src/main/java/io/dropwizard/redis/RedisClientBundle.java#L29) on startup.

Wouldn't a cleaner/simpler solution to be to override the `run` method of `RedisClientBundle`  to make it catch connection exceptions during startup and switch to a 'disabled/not available' mode? That way we can continue to use the maintained and instrumented official dropwizard client.

Reply:
Yes I should've made it clearer in the ADR but the line of code (`redisClient.connect(codec)`) will throw an exception if it can't connect to Redis.

Reply:
I think overriding the `run` method and switching to a 'disabled/not available' mode makes things unnecessarily complicated because we'd still need to manage the `RedisClientManager` (see https://github.com/dropwizard/dropwizard-redis/blob/master/src/main/java/io/dropwizard/redis/RedisClientFactory.java#L57) but we can't manage it if we don't have a `connection`, and we can't have a `connection` if a connection to Redis can't be made. Managing the lifecycle  is part of the point of using the dropwziard-redis library. 
As a workaround we've already built a custom RedisClientManager (see https://github.com/alphagov/pay-publicapi/blob/master/src/main/java/uk/gov/pay/api/managed/RedisClientManager.java) which doesn't require a connection and this code isn't much extra effort.

Reply:
I've updated the point to say `redisClient.connect(codec)` will throw an exception if it can't connect to Redis and thus cause the application to fail to start up.